### ISSUE TITLE
write_image hangs

### ISSUE THREAD

**Issue by neuron-whisperer** [plotly/Kaleido#134]
I'm running Python 3.9.6 on Windows 10 with Plotly 5.7.0 and Kaleido 0.2.1.

When I call:

    figure.write_image(buffer)

...Plotly/Kaleido never returns and cannot be gracefully terminated; the script must be force-closed.

I first noticed this issue a year ago. The suggestions at that time were ineffective. Here we are a year later, at least two other users have posted the same issue, and there has been no resolution.

**Comment by jakobbrown** [comment]
I have the same issue. Kernel unable to be interrupted once called and never executes. Same for fig.to_image(), and with any image format. this is using Plotly 5.4.0 and Kaleido 0.2.1 in a conda environment on windows 10. Unable to get a keyboard interrupt traceback as keyboard interrupt doesn't work.

**Comment by gygabyte017** [comment]
Same here, last year with Python 3.8 and Plotly 4.14.3 it seemed to become stable somehow but never figured out why and how, now just upgraded to Python 3.9 and Plotly 5.7.0 as per my project requirements, and write_image randomly hangs with the exact same code in the exact same machine/configuration.

**Comment by Brainor** [comment]
> Same here, last year with Python 3.8 and Plotly 4.14.3 it seemed to become stable somehow but never figured out why and how, now just upgraded to Python 3.9 and Plotly 5.7.0 as per my project requirements, and write_image randomly hangs with the exact same code in the exact same machine/configuration.

Yes, the problem is that with same code and same environment, it hangs randomly, which doesn't make any sense.  
In my case, I use gnu-parallel to do the exactly same jobs to produce .png files(with different input of course and same code), but some jobs produce image file while some don't and just hang indefinitely. So it has to be memory allocation or IO problem I assume.

**Comment by sirtyman** [comment]
The same issue happens in my environment. I've tried different Python versions (3.8 - 3.10). It seems that the problem happens on 64 bit version only.

**Comment by MahdiDavari** [comment]
a quick temporary fix is to use an old version before they fix it!
This version worked for me:
pip install kaleido==0.2.1.post1

**Comment by c5creative** [comment]
In our case, we had to revert to `pip install kaleido==0.1.0.post1` in win10/Python3.9

**Comment by ukewea** [comment]
> In our case, we had to revert to `pip install kaleido==0.1.0.post1` in win10/Python3.9

Thanks for the hint, my `write_image()` call hangs as well in Windows 11/Python 3.10 before reverting to `kaleido==0.1.0.post1`

**Comment by cmosguy** [comment]
Not sure what @MahdiDavari was referring too but `0.2.1.post1` is not a legit package:

```
$ pip install kaleido==0.2.1.post1
WARNING: Ignoring invalid distribution -orch (c:\users\kleinada\anaconda3\envs\ds\lib\site-packages)
WARNING: Ignoring invalid distribution -orch (c:\users\kleinada\anaconda3\envs\ds\lib\site-packages)
ERROR: Could not find a version that satisfies the requirement kaleido==0.2.1.post1 (from versions: 0.0.1rc3, 0.0.1rc4, 0.0.1rc5, 0.0.1rc6, 0.0.1rc8, 0.0.1rc9, 0.0.1, 0.0.2, 0.0.3, 0.0.3.post1, 0.1.0a2, 0.1.0a3, 0.1.0, 0.1.0.post1, 0.2.0rc1, 0.2.0, 0.2.1)
ERROR: No matching distribution found for kaleido==0.2.1.post1
```
probably have to stick with `kaleido==0.1.0.post1`

**Comment by chenyhsu** [comment]
Same on archlinux.

python 3.10.10
kaleido 0.2.1
plotly 5.14.1

It always success for first run, but if create image multiple times in a loop, it would randomly hangs.

CPU usage is high, but there is no error message.

**Comment by jgraber** [comment]
I have the same blocking behaviour with this combination:

Windows: 10 22H2
Python: 3.11.3 (64bit)
plotly: 5.14.1
kaleido: 0.2.1

**Comment by florian6973** [comment]
Same issue with Windows11, plotly 5.14.1, kaleido: 0.2.1

**Comment by howick19** [comment]
I have the same issue with the ToImage() function in plotly.Blazor.  Works if the plot has no data, but as soon as data is added it hangs.

**Comment by ssabdb** [comment]
After having had this happen to me multiple times - It looks to me like this is happening in with some sort of subprocess communication happening with the kaleido subprocess, lines 300-308. 

I'm using Python 3.10.11, GCC 11.2.0 on RHEL. kaleido: 0.2.1. This happened eventually to 2 subprocesses running in a loop. 

The stack traces to me cancelling the process indicate where it is getting stuck in the same place in two completely independent processes... 

in scopes scopes/base.py:
Line 308, 
`response = self._proc.stdout.readline()`

I've not spent a lot of time working through this unfortunately, but for whatever reason the Kaleido C++/python communication is getting stuck on waiting for a response. 

For anyone using plotly switching to the Orca engine works in my environment, but is slow.

**Comment by gioxc88** [comment]
it's incredible this hasn't been fixed yet

**Comment by NBardelot** [comment]
+1

**Comment by hmgaudecker** [comment]
In a [course](https://effective-programming-practices.vercel.app/landing-page.html) teaching programming to economists we have switched to plotly a while back. We were unaware of this issue back then because all course developers are on Linux or Mac.

For better or worse, for Masters' theses and papers we are stuck with static images. In a small block course we gave today, 5 out of 6 Windows users could not export their images.  I really don't want to go back to matplotlib + seaborn, but explaining that to 50+ students in the upcoming term would also not be great. 

The number may be [small relative to total usage](https://github.com/plotly/Kaleido/issues/150#issuecomment-1293508017), but for us it is very large. Would there be any updates to the status report a year ago by any chance, @nicolaskruchten?

And btw, thanks a lot for a great tool (i.e., plotly) that I am  enjoying everyday in notebooks!!!

**Comment by gioxc88** [comment]
> In a [course](https://effective-programming-practices.vercel.app/landing-page.html) teaching programming to economists we have switched to plotly a while back. We were unaware of this issue back then because all course developers are on Linux or Mac.
> 
> 
> 
> For better or worse, for Masters' theses and papers we are stuck with static images. In a small block course we gave today, 5 out of 6 Windows users could not export their images.  I really don't want to go back to matplotlib + seaborn, but explaining that to 50+ students in the upcoming term would also not be great. 
> 
> 
> 
> The number may be [small relative to total usage](https://github.com/plotly/Kaleido/issues/150#issuecomment-1293508017), but for us it is very large. Would there be any updates to the status report a year ago by any chance, @nicolaskruchten?
> 
> 
> 
> And btw, thanks a lot for a great tool (i.e., plotly) that I am  enjoying everyday in notebooks!!! 

If you follow the above comment and install the write version it should still work👍

**Comment by hmgaudecker** [comment]
You mean installing `kaleido==0.1.0.post1` ?

**Comment by gioxc88** [comment]
> You mean installing `kaleido==0.1.0.post1` ? 

Yes, that fixes for me

**Comment by ebongo1** [comment]
I also had this same issue using Python 3.11.3 on a 64 bit Windows machine, and installing kaleido==0.1.0.post1 worked. However, my teammate is trying to install that kaleido version on his Linux machine and is getting "ERROR: No matching distribution found for kaleido==0.1.0.post1". But when he installed the latest version of kaleido (0.2.1) things worked fine for him and it didn't "hang" on the write_image() function like it does on mine with the latest version.

**Comment by OverLordGoldDragon** [comment]
If it's not getting fixed, a warning should be thrown. https://github.com/plotly/plotly.py/issues/4401

Note, `conda install -c conda-forge python-kaleido==0.1.0` also worked for me.

**Comment by hmgaudecker** [comment]
So it seems like the above workaround does help. But it is the **only** thing for preventing cross-platform environment files (provided there are packages for all platforms, of course), since that lone working version (Windows) does not exist on other platforms:

```shell-session
$ pip install kaleido==0.1.0.post1
ERROR: Could not find a version that satisfies the requirement kaleido==0.1.0.post1 (from versions: 0.0.1rc3, 0.0.1rc4, 0.0.1rc5, 0.0.1rc6, 0.0.1rc8, 0.0.1rc9, 0.0.1, 0.0.2, 0.0.3, 0.0.3.post1, 0.1.0a2, 0.1.0a3, 0.1.0, 0.2.0rc1, 0.2.0, 0.2.1)
ERROR: No matching distribution found for kaleido==0.1.0.post1
```

I understand that it might be too difficult to come up with a new version of kaleido that would actually solve the problems. Just re-releasing that old thing on Windows or at least the warning suggested in https://github.com/plotly/plotly.py/issues/4401 with clear instructions on what to do would help a ton!

**Comment by CommanderPho** [comment]
> it's incredible this hasn't been fixed yet

I agree completely, the `0.1.0.post1` version workaround worked for me on Win10 Python 3.9 but this functionality (saving a produced image) just seems to foundational that I don't understand why it's not a top priority. It only takes one incident like this before potential plotly users are turned off from ever trying it again.

**Comment by gza-austin-elliott** [comment]
Datapoint: kaleido 0.2.1 hangs on my environment. Work pushed us from conda and was previously using orca through conda.

Running Windows 10.0.19045, VSC, Python 3.12
Regular install of kaleido hung on write_image(), same as above. Removal was even difficult from my venv, pip uninstall didn't remove from site-packages. Lib was engaged when I tried to delete, found kaleido.exe running in the background via Task Manager but ending that still did not disengage whatever was running in the background. No soft reboots of apps fixed.

Restart computer and all apps. Was able to delete manually package from \venv\site-packages\. Installed `pip install kaleido==0.1.0.post1` and write_image() is working as expected.

**Comment by Lenakeiz** [comment]
I had the same problem using python 3.10 and going back to the suggested kaleido==0.1.0.post1 worked fine for me. Thanks to whoever suggested that!

**Comment by maxmax91** [comment]
Same issue with win11, downgraded and it works. Any further information? Any way to fix it?

**Comment by neuropathbasel** [comment]
Yes, there is a fix which I finally found today: https://github.com/plotly/Kaleido/issues/110#issuecomment-2113369864

**Comment by gvwilson** [comment]
Thanks for your interest in Kaleido. We are currently working on an overhaul that might address your issue - we hope to have news in a few weeks and will post an update then. Thanks - @gvwilson

**Comment by aym-0401** [comment]
Please fix this. If I had known about this issue I wouldn't have written a month worth of code with the Plotly library. This small issue is wasting tons of coding efforts I've made.

**Comment by JonNir1** [comment]
Hi, Not sure if there's a deadline to fix this, but I am hoping it's soon :)
In any case, it's probably best to add a warning as [suggested here](https://github.com/plotly/plotly.py/issues/4401), @gvwilson

**Comment by Tendx** [comment]
> The same issue happens in my environment. I've tried different Python versions (3.8 - 3.10). It seems that the problem happens on 64 bit version only.

I'm used to using Anaconda Envienment. Finally, I installed a last x86 anaconda package **Anaconda3-2022.05-Windows-x86.exe**, and use install kaleido 0.2.1 by pip. It works!
THANK YOU : )

**Comment by ThierryBotT** [comment]
> In our case, we had to revert to `pip install kaleido==0.1.0.post1` in win10/Python3.9

Did not work for me, it hangs also. :(

**Comment by ayjayt** [comment]
@tillschwoerer there had been a bug in the plotly --> kaleido wrapper in the very first release, so I'm curious as to what versions (kaleido, plotly) were hanging.

For everybody else:

v1.0.0 of kaleido is currently in prerelease with a significant redesign. Plotly versions 6.1.1 and greater are compatible with kaleido v1. Upgrading to kaleido==1.0.0rc13 and Plotly==6.1.1 may resolve all the issues. Plotly's early june release will use kaleido>=1.0.0 by default.

**Comment by tillschwoerer** [comment]
@ayjayt Your answer explains why it did not work in my case. I was running kaleido==1.0.0rc13 and Plotly==6.0.1. I just upgrade Plotly to 6.1.1 and it works fine. Excellent!

**Comment by gvwilson** [comment]
thank you for letting us know - we're glad it's working
